File: 04-implementation/backend/app/models/programme.py
from datetime import datetime
from sqlalchemy import String, Integer, Float, Text, ForeignKey, DateTime, Enum
from sqlalchemy.orm import Mapped, mapped_column, relationship
import enum

from app.database import Base


class CourseCategory(str, enum.Enum):
    major = "major"           # หมวดวิชาเฉพาะ
    general = "general"       # หมวดวิชาศึกษาทั่วไป
    elective = "elective"     # หมวดวิชาเลือกเสรี


class Programme(Base):
    __tablename__ = "programmes"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    university: Mapped[str] = mapped_column(String(255), nullable=False)
    degree: Mapped[str] = mapped_column(String(100), nullable=True)
    tqf_filename: Mapped[str] = mapped_column(String(512), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    extraction_status: Mapped[str] = mapped_column(String(50), default="pending")

    courses: Mapped[list["Course"]] = relationship(back_populates="programme", cascade="all, delete-orphan")


class Course(Base):
    __tablename__ = "courses"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    programme_id: Mapped[int] = mapped_column(ForeignKey("programmes.id"), nullable=False)
    code: Mapped[str] = mapped_column(String(20), nullable=True)
    name_th: Mapped[str] = mapped_column(String(512), nullable=True)
    name_en: Mapped[str] = mapped_column(String(512), nullable=True)
    description_th: Mapped[str] = mapped_column(Text, nullable=True)
    description_en: Mapped[str] = mapped_column(Text, nullable=True)
    credits: Mapped[float] = mapped_column(Float, default=3.0)
    category: Mapped[CourseCategory] = mapped_column(Enum(CourseCategory), default=CourseCategory.major)

    programme: Mapped["Programme"] = relationship(back_populates="courses")
    skills: Mapped[list["CourseSkill"]] = relationship(back_populates="course", cascade="all, delete-orphan")

    @property
    def credit_weight(self) -> float:
        weights = {CourseCategory.major: 1.0, CourseCategory.general: 0.5, CourseCategory.elective: 0.0}
        return self.credits * weights.get(self.category, 1.0)


class CourseSkill(Base):
    __tablename__ = "course_skills"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    course_id: Mapped[int] = mapped_column(ForeignKey("courses.id"), nullable=False)
    raw_skill: Mapped[str] = mapped_column(String(512), nullable=False)
    cluster_id: Mapped[int] = mapped_column(ForeignKey("skill_clusters.id"), nullable=True)

    course: Mapped["Course"] = relationship(back_populates="skills")

File: 04-implementation/backend-rust/shared/src/models/programme.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Programme {
    pub id: i32,
    pub name: String,
    pub university: String,
    pub degree: Option<String>,
    pub tqf_filename: Option<String>,
    pub created_at: NaiveDateTime,
    pub extraction_status: String,
    pub extraction_error: Option<String>,
    pub extracted_at: Option<NaiveDateTime>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Course {
    pub id: i32,
    pub programme_id: i32,
    pub code: Option<String>,
    pub name_th: Option<String>,
    pub name_en: Option<String>,
    pub description_th: Option<String>,
    pub description_en: Option<String>,
    pub credits: f64,
    pub category: String,
    pub updated_at: NaiveDateTime,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct CourseSkill {
    pub id: i32,
    pub course_id: i32,
    pub skill_term: String,
    pub source: String,
    pub cluster_id: Option<i32>,
}
